posts/2022/02/making-gpus-accessible-to-pydata-ecosystem-via-array-api.md

```diff
@@ -0,0 +1,484 @@
+library we'll use CuPy, which is a NumPy/SciPy-compatible array library for
+GPU-accelerated computing with Python.
+
+#### Changes to SciPy, scikit-learn, scikit-image, CuPy
+
+Since SciPy, scikit-learn, scikit-image are designed mainly for NumPy, we need to use Array API
+to dispatch to CuPy or NumPy based on the input. We do that by changing all the functions
+that are called in the above demo to use `get_namespace` to replace all the instances of
+`np.` (or `numpy.`) to properly dispatch to the appropriate array library. This is straightforward
+at times and not so trivial at many. We'll see that in the challenges in next section.
+
+The final code for the demo can be see [here](https://github.com/Quansight-Labs/amd-demo/blob/master/plot_coin_segmentation.ipynb).
+
+## Challenges and Workarounds
+
+At the moment Array API implements plenty of the commonly used functions from the NumPy API,
+which makes it easier to replace the NumPy API functions with those, although there are a
+bunch of rough edges, which makes the process a bit harder. These are the three main issues, we faced
+in the process:
+
+#### 1. Compiled code
+
+There is a lot of functionality in SciPy and Scikits that is not pure
+  Python, critical to performance parts are written using compiled code extensions. These parts
+  are dependent on NumPy's internal memory representation, which makes them CPU only. This
+  restricts the usage Array API to dispatch such code. There is another project `uarray`
+  for dispatching compiled code, which needs to be used along with Array API dispatching to make
+  it fully generic.
+
+See for instance following piece of code:
+
+```python
+from scipy import sparse
+
+graph = sparse.coo_matrix(<input_arrays>)
+```
+
+Here the `sparse.coo_matrix` is a compiled code extension in the `scipy.sparse` module. We need the
+dispatch to `cupyx.scipy.sparse.coo_matrix` for CuPy arrays and `scipy.sparse.coo_matrix` for NumPy
+arrays. This is not possible with the Array API's dispatching via `get_namespace`. For this demo we have
+added a workaround, until we have implemented the support for `uarray` based backed in SciPy. The
+workaround looks something like this for CuPy:
+
+```python
+xp, _ = get_namespace(<input_arrays>)
+
+if xp.__name__ == 'cupy.array_api':
+  from cupyx.scipy import sparse as cupy_sparse
+  coo_matrix =  cupy_sparse.coo_matrix
+else:
+  coo_matrix = scipy.sparse.coo_matrix
+
+graph = coo_matrix(<input_arrays>)
+```
+
+Note that, this workaround is only valid for NumPy/CuPy arrays.
+
+
+#### 2. Not implemented functions
+
+There are still a bunch of functions in the NumPy API which don't
+have a direct equivalent in the Array API, either due to the fact that those functions are not
+generic enough to be defined in the Array API or we haven't got around implementing those.
+
+Some examples of unimplemented functions that we encountered during this excercise:
+
+
+* `np.atleast_3d`
+* `np.strides`
+* `np.real`
+* `np.clip`
+
```

Comment:
Here we need a line concluding what was done to resolve the challenge. Was it a matter of rewriting the code in each case @aktech?

Reply:
For some of them it was reimplementing an equivalent function and for remaining it was a hack basically, which is calling those functions on `._array` attribute and then converting back to array api.